clippy_lints/src/incorrect_impls.rs

```diff
@@ -84,7 +97,36 @@ impl LateLintPass<'_> for IncorrectImpls {
                     copy_def_id,
                     trait_impl.substs,
                 )
+            && let ty::Adt(def, substs) = hir_ty_to_ty(cx.tcx, imp.self_ty).kind()
         {
+            let fields = def.all_fields().collect_vec();
+            // Necessary as `all` returns true on an empty iterator
+            if !fields.is_empty()
+                && fields.iter().all(|field| {
+                    let ty = field.ty(cx.tcx, substs);
+
+                    match ty.kind() {
+                        // `MaybeUninit<T>`
+                        ty::Adt(_, _) if is_type_lang_item(cx, ty, LangItem::MaybeUninit) => true,
+                        // `[MaybeUninit<T>; N]`
+                        ty::Array(inner_ty, _) | ty::Slice(inner_ty)
+                            if is_type_lang_item(cx, *inner_ty, LangItem::MaybeUninit) =>
+                        {
+                            true
+                        },
+                        // Other cases are likely pretty rare.
+                        _ => false,
+                    }
```

Comment:
```suggestion
                    match ty.kind() {
                        // `MaybeUninit<T>`
                        ty::Adt(_, _) => is_type_lang_item(cx, ty, LangItem::MaybeUninit),
                        // `[MaybeUninit<T>; N]`
                        ty::Array(inner_ty, _) | ty::Slice(inner_ty) => is_type_lang_item(cx, *inner_ty, 
                        // Other cases are likely pretty rare.
                        _ => false,
                    }
```

Reply:
I believe there was some additional work done in the block before, but I'll change it :+1: